src/stdio_mgr/types.py

```diff
@@ -0,0 +1,175 @@
+    the context.
+
+    Used as a context manager, it will close all of the streams on exit.
+    however it does not open the streams on entering the context manager.
+
+    No exception handling is performed while closing the streams, so any
+    exception occurring the close of any stream renders them all in an
+    unpredictable state.
+    """
+
+    def __new__(cls, iterable):
+        """Instantiate new tuple from iterable containing three streams."""
+        items = list(iterable)
+        assert len(items) == 3, "iterable must be three items"  # noqa: S101
+
+        return super(StdioTupleBase, cls).__new__(cls, items)
+
+    @property
+    def stdin(self):
+        """Return stdin stream."""
+        return self[0]
+
+    @property
+    def stdout(self):
+        """Return stdout stream."""
+        return self[1]
+
+    @property
+    def stderr(self):
+        """Return stderr stream."""
+        return self[2]
+
+
+class ClosingStdioTuple(StdioTupleBase):
+    """Tuple context manager of stdin, stdout and stderr stream-like objects.
+
+    Used as a context manager, it will close all of the streams on exit.
+    however it does not open the streams on entering the context manager.
+
+    No exception handling is performed while closing the streams, so any
+    exception occurring the close of any stream renders them all in an
+    unpredictable state.
+    """
+
+    def close(self):
+        """Close all streams."""
+        return list(self._map("close"))
+
+    def safe_close(self):
+        """Close all streams, ignoring any ValueError."""
+        return list(self.suppress_map(ValueError, "close"))
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        """Exit context, closing all members."""
+        self.close()
+        return super().__exit__(exc_type, exc_value, traceback)
+
+
+class TextIOTuple(StdioTupleBase):
+    """Tuple context manager of stdin, stdout and stderr TextIOBase objects."""
+
+    _ITEM_BASE = TextIOBase
+
+    # pytest and colorama inject objects into sys.std* that are not real TextIOBase
+    # and fail the assertion of this class
+
+    def __new__(cls, iterable):
+        """Instantiate new tuple from iterable containing three TextIOBase streams."""
+        self = super(TextIOTuple, cls).__new__(cls, iterable)
+        if not self._all(lambda item: isinstance(item, cls._ITEM_BASE)):
+            raise ValueError(
+                "iterable must contain only {}".format(cls._ITEM_BASE.__name__)
+            )
+        return self
+
+
+class FakeIOTuple(StdioTupleBase):
+    """Tuple context manager of stdin, stdout and stderr-like objects."""
+
+
+class AnyIOTuple(StdioTupleBase):
+    """Tuple context manager which will create FakeIOTuple or TextIOTuple."""
+
+    def __new__(cls, iterable):
+        """Instantiate new TextIOTuple or FakeIOTuple from iterable."""
+        items = list(iterable)
+        if any(not isinstance(item, TextIOTuple._ITEM_BASE) for item in items):
```

Comment:
If the type check in `TextIOTuple.__new__()` were stronger, this could just be:

```
try:
    return TextIOTuple(items)
except {{{RelevantError}}}:
    return FakeIOTuple(items)
```